Numark Mixtrack (Pro) 3 FX mapping change (V1.7)

djsteph opened the pull request:
Code changes have been validated with JSHint.com

**Bug fixes**
Touch platter + wheel touch speeds up/slows down song when wheel not enabled 
Corrected Tap button mapping
Correct Tap button led behavior when shift lock is active
Missing callback function for smartPFL to work properly if track is loaded directly on UI
Adjusted mapping of AutoLoop to be identical to Numark user manual, Serato and Virtual DJ, 
(1, 2, 4, 8) instead of 2, 4, 8, 16 if not shifted and 0.125, 0.25, 0.5, 16 if shifted

**FX section Mapping changes**
Added InstantFx feature: FX are added to InstantFx by holding Tap and FX button. InstantFx are enabled/disabled directly with the Strip.

Tap + FX button: Enables / Disables InstantFx.

Strip: Super effect. If Instant FX is activated for an FX, also enables / disables that FX
Shift + Strip: Fast Seek (already mapped)
Tap button + Strip FX paramerer 1 - Effects 1, 2 and 3

Beat knob: Mix/ratio
TAP + Beat knob = parameter 2 of FX 1, 2 and 3
Shift + Beat knob: Beat grid move
Pad Mode + Beat knob: Sampler volumes (deck 1= 1-4; deck 2 = (5-8)

Tap Button: well.. Tap button.

**Deck 1:** 
SHIFT + Treble Knob = FX paramerer 1 - Effect 1
SHIFT + Medium Knob = FX parameter 2 - Effect 1
SHIFT + Bass Knob = Fx parameter 3 - Effect 1
SHIFT + Filter Knob = Fx parameter 4 - Effect 1

**Deck 2:** 
SHIFT + Treble Knob = FX paramerer 1 - Effect 2
SHIFT + Medium Knob = FX parameter 2 - Effect 2
SHIFT + Bass Knob = Fx parameter 3 - Effect 2
SHIFT + Filter Knob = Fx parameter 4 - Effect 2

**Deck 1:** 
Pad Mode + Treble Knob = FX paramerer 1 - Effect 3
Pad Mode + Medium Knob = FX parameter 2 - Effect 3
Pad Mode + Bass Knob = Fx parameter 3 - Effect 3
Pad Mode + Filter Knob = Fx parameter 4 - Effect 3

**Deck 2:** 
Pad Mode + Treble Knob = FX paramerer 1 - Effect 4
Pad Mode + Medium Knob = FX parameter 2 - Effect 4
Pad Mode + Bass Knob = Fx parameter 3 - Effect 4
Pad Mode + Filter Knob = Fx parameter 4 - Effect 4

**Other changes**
Connected controls for FX enabled 
Removed soft takeover for Super FX knob (accessed by strip)
Removed Callback function for FX Loaded to prevent all parameters from being automatically linked to Super Knob
Removed configuration option BeatKnobAsSamplerVolume
Included Parameter soft taker over function

djsteph committed:
Merge remote-tracking branch 'refs/remotes/upstream/master' into Mixtrack-Pro-3-mapping-v1.6

djsteph committed:
Numark Mixtrack (Pro) 3 FX mapping change (V1.7)

**Bug fixes **
Touch platter + wheel touch speeds up/slows down song when wheel not
enabled
Corrected Tap button mapping
Correct Tap button led behavior when shift lock is active

Adjusted mapping of AutoLoop to be identical to Numark user manual,
Serato and Virtual DJ,
(1, 2, 4, 8) instead of 2, 4, 8, 16 if not shifted and 0.125, 0.25, 0.5,
16 if shifted

**FX section Mapping changes**

Added InstantFx feature: FX are added to InstantFx by holding Tap and FX
button. InstantFx are enabled/disabled directly with the Strip.

Tap + FX button: Enables / Disables InstantFx.

Strip = Parameter 1 of FXs,
TAP + Strip = Super knob

Beat knob = parameter 2 of FXs,
TAP + Beat knob = Mix ratio
PadModeButton + Beat knob = Sampler volumes
Shift + BeatKnob = Beat grid move

Connected controls for FX enabled
Removed soft takeover for Super FX knob (accessed by strip)
Removed Callback function for FX Loaded to prevent all parameters from
being automatically linked to Super Knob
Removed configuration option BeatKnobAsSamplerVolume

djsteph committed:
Bug fix -InstantFX handling with Shift Key

Bug fix -InstantFX handling with Shift Key

illuusio commented:
You should clean that 'Merge remote-tracking branch 'refs/remotes/upstream/master' into Mixtrack-Pro-3-mapping-v1.6' away from commits

illuusio commented:
Have you run static checking tools against this like it's recommended in documentation?

Be-ing commented:
From the [forum thread](http://mixxx.org/forums/viewtopic.php?f=7&t=7286&start=180) it looks like the functionality is still being decided upon. Shall I wait until you've settled on that for a thorough review?

djsteph commented:
Agreed. What is currently being decided on is the additional mapping of FX Parameter control. It does add a bit of code to the file, so I will have to clean up that code also. It can be another commit on this pull request, or a new pull request all together. 

My GitHub skills being weak, I have no idea on how to do the following: 

> You should clean that 'Merge remote-tracking branch 'refs/remotes/upstream/master' into Mixtrack-Pro-3-mapping-v1.6' away from commits

Be-ing commented:
> It can be another commit on this pull request, or a new pull request all together.

No need to open another PR. Just add commit(s) to this branch on your computer, push to this branch of your fork on GitHub, and they will appear on this PR.

> My GitHub skills being weak, I have no idea on how to do the following

Neither do. @illuusio , could you explain or provide a link that explains?

djsteph committed:
Add FX Parameter mapping

Add FX Parameter mapping and correct smartPFL callback

New mapping:
Deck 1:
SHIFT + Treble Knob = FX paramerer 1 - Effect 1
SHIFT + Medium Knob = FX parameter 2 - Effect 1
SHIFT + Bass Knob = Fx parameter 3 - Effect 1
SHIFT + Filter Knob = Fx parameter 4 - Effect 1

Deck 2:
SHIFT + Treble Knob = FX paramerer 1 - Effect 2
SHIFT + Medium Knob = FX parameter 2 - Effect 2
SHIFT + Bass Knob = Fx parameter 3 - Effect 2
SHIFT + Filter Knob = Fx parameter 4 - Effect 2

Deck 1:
Padmode + Treble Knob = FX paramerer 1 - Effect 3
Padmode + Medium Knob = FX parameter 2 - Effect 3
Padmode + Bass Knob = Fx parameter 3 - Effect 3
Padmode + Filter Knob = Fx parameter 4 - Effect 3

Deck 2:
Padmode + Treble Knob = FX paramerer 1 - Effect 4
Padmode + Medium Knob = FX parameter 2 - Effect 4
Padmode + Bass Knob = Fx parameter 3 - Effect 4
Padmode + Filter Knob = Fx parameter 4 - Effect 4

djsteph committed:
Include updated XML file for version 1.8

file updates forgotten in previous commit

djsteph committed:
Eliminate unnecessary object "OnFXButtonControl"

Eliminate unnecessary object "OnFXButtonControl"

djsteph commented:
@Be-ing , I am confident that you can start reviewing the code. I made final commits and adjusted initial pull request description to reflect all changes made in the past few days.

sblaisot commented on res/controllers/Numark-Mixtrack-3.midi.xml:
```diff
@@ -1,12 +1,12 @@
 	<description>The Numark Mixtrack 3 and Numark Mixtrack Pro 3 are the same controller except that the Pro version has an integrated sound card.</description>
         <forums>http://mixxx.org/forums/viewtopic.php?f=7&amp;t=7286</forums>
-        <wiki>http://www.mixxx.org/wiki/doku.php/numark_mixtrack_pro_3</wiki>
+	    <wiki>http://www.mixxx.org/wiki/doku.php/numark_mixtrack_pro_3</wiki>
```
tab indentation -> spaces

sblaisot commented on res/controllers/Numark-Mixtrack-3.midi.xml:
```diff
@@ -895,6 +895,26 @@
+                    <script-binding/>
+                </options>
+            </control>			
+			<control>
```
indentation problem

sblaisot commented on res/controllers/Numark-Mixtrack-3.midi.xml:
```diff
@@ -895,6 +895,26 @@
 			<control>
                 <group>[Channel1]</group>
+                <key>NumarkMixtrack3.InstantFXOff</key>
+				<description>Channel 1 strip Instant FX Off</description>
```
indentation problem

sblaisot commented on res/controllers/Numark-Mixtrack-3.midi.xml:
```diff
@@ -895,6 +895,26 @@
+            <control>
+                <group>[Channel2]</group>
+                <key>NumarkMixtrack3.InstantFXOff</key>
+				<description>Channel 2 strip Instant FX Off</description>
```
indentation problem

djsteph replied:
corrected

sblaisot commented:
There is a mix of tab and space indentation.
Please use jsbeautifier to update all indentation problems at once (and verify coding style in general).

sblaisot commented on res/controllers/Numark-Mixtrack-3-scripts.js:
```diff
@@ -1984,138 +2119,245 @@ NumarkMixtrack3.PitchBendPlusButton = function(channel, control, value, status,
+    
     if (value===DOWN) {
+        deck.TapDown = true;
+        bpm.tapButton(decknum);
```
isn't bpm an undefined variable ?

djsteph replied:
bpm is from the common controller script. Do I need to redefine it here also?

djsteph replied:
Done

sblaisot commented on res/controllers/Numark-Mixtrack-3-scripts.js:
```diff
@@ -2227,6 +2469,34 @@ NumarkMixtrack3.OnPFLStatusChange = function(value, group, control) {
+            engine.setValue(group, 'pfl', true);
+            engine.setValue(groupOff, 'pfl', false);
+            deck.LEDs["headphones"].onOff(ON);
+            deckOff.LEDs["headphones"].onOff(ON);
```
shouldn't the second be OFF ?

djsteph replied:
The headphones LED is set on one channel. When you load on the other deck, the LED switches to the opposite state with that code. I tested physically on the controller, and LED behavior is good.

djsteph replied:
Fixed

sblaisot commented:
You can probably merge NumarkMixtrack3.TrebleKnob NumarkMixtrack3.MiddleKnob and NumarkMixtrack3.BassKnob into one function to avoid code duplication.

sblaisot commented on res/controllers/Numark-Mixtrack-3-scripts.js:
```diff
@@ -148,25 +149,44 @@ var intervalsPerRev = 1200,
- *            - Add script to control "volume","filterHigh","filterMid","filterLow" and"crossfader"
- *              In order for engine.softTakeover to work for these controls
+ *            - Add script to control "volume" and"crossfader"
+ *              in order for engine.softTakeover to work for these controls
```
You can use soft takeover in pure XML: 
http://www.mixxx.org/wiki/doku.php/midi_controller_mapping_file_format#input_options
use Soft-takeover modifier

djsteph replied:
Ok, noted

djsteph replied:
> You can probably merge NumarkMixtrack3.TrebleKnob NumarkMixtrack3.MiddleKnob and NumarkMixtrack3.BassKnob into one function to avoid code duplication

How would you do it?

sblaisot replied:
You can use the "control" parameter which handles the midino of the control to select which parameter you move.

djsteph replied:
Done, I had to include a switch in the function to manage the parameter numbers for EQ and FX has they were not sequencial

sblaisot commented on res/controllers/Numark-Mixtrack-3-scripts.js:
```diff
@@ -331,6 +351,20 @@ function sendShortMsg(control,midino,value) {
     midi.sendShortMsg(control, midino, value);
 }
 
+function parameterSoftTakeOver (group, control, value) {
```
Why don't you use the built-in soft-takeover ?
http://www.mixxx.org/wiki/doku.php/midi_scripting#soft-takeover

djsteph replied:
The soft take over in Mixxx only works with engine.setvalue, not with engine.setparameter. There is a documented request in the launchpad

sblaisot replied:
O.O Oh ! Right.

Be-ing replied:
Augh. This bug should really be fixed for 2.1...

sblaisot commented on res/controllers/Numark-Mixtrack-3-scripts.js:
```diff
@@ -1036,8 +1087,8 @@ NumarkMixtrack3.initButtonsObjects = function() {
-            new SingleDoubleBtn(NumarkMixtrack3.OnShiftedPFLButton);
-    }
+            new SingleDoubleBtn(NumarkMixtrack3.OnShiftedPFLButton);       
+    }    
```
extra spaces at end of line ?

djsteph replied:
must have been a tab that Notepad++ converted

djsteph replied:
corrected

Be-ing replied:
I don't know about Notepad++, but KWrite/Kate/KDevelop have an option to show dots to indicate trailing spaces. That helps with noticing them before git does.

djsteph replied:
Thanks for this tip! I found the option in Notepad++ to show spaces, tabs and all characters

sblaisot commented on res/controllers/Numark-Mixtrack-3-scripts.js:
```diff
@@ -1104,23 +1155,19 @@ NumarkMixtrack3.init = function(id,debug) {
+        
     } 
-
+   /* 
```
this comment becomes a single-line comment, so please switch to //

djsteph replied:
Removed

sblaisot commented on res/controllers/Numark-Mixtrack-3-scripts.js:
```diff
@@ -1164,6 +1211,11 @@ NumarkMixtrack3.init = function(id,debug) {
+    
+    // The duration control is being used to tell if a track has successfully loaded for smartPFL feature
+    engine.connectControl("[Channel1]", "duration", "NumarkMixtrack3.OnLoadSelectedTrack");
+    engine.connectControl("[Channel2]", "duration", "NumarkMixtrack3.OnLoadSelectedTrack");
```
You should use /track_samples/ instead of /duration/ as duration can change when you adjust pitch ubless you want to activate PFL on pitch adjustment.

djsteph replied:
I just tried that switch to use track_samples and the feature stopped working. I also tested modifying the pitch while the other track had PFL on, and no adverse effects occured (activate PFL on pitch adjustment.)

illuusio commented:
Actually rethinking this because there is reviews and stuff. Intrusive git-fu (https://sethrobertson.github.io/GitFixUm/fixup.html) changes thing so drastically that it'll make everyone unhappy. So we have three options neater close this or cherry-pick good commits by hand by someone who has commits rights to Mixxx or we can trust git it solves this as it says: this branch has no conflicts.

djsteph reviewed (commented):
I'll clean up the spaces & tabs... Please see my replies to comments in code for explanations.

Be-ing commented:
It would be nice if we could integrate JSHint into Travis.

Be-ing commented on res/controllers/Numark-Mixtrack-3-scripts.js:
```diff
@@ -1984,138 +2119,245 @@ NumarkMixtrack3.PitchBendPlusButton = function(channel, control, value, status,
+    var deck = NumarkMixtrack3.decks["D" + decknum];
+    
+    if(!deck.shiftKey && !deck.PADMode ) {
+    parameterSoftTakeOver ("[EqualizerRack1_[Channel"+ [decknum] +"]_Effect1]", "parameter3", value); 
```
Why is `decknum` in brackets here?

djsteph replied:
corrected... not a pro in javascript

Be-ing commented on res/controllers/Numark-Mixtrack-3-scripts.js:
```diff
@@ -1984,138 +2119,245 @@ NumarkMixtrack3.PitchBendPlusButton = function(channel, control, value, status,
+    }
+};
+
+//parameterSoftTakeOver (group, control, value) 
```
This looks like experimental code that you forgot to remove.

djsteph replied:
Removed

Be-ing commented on res/controllers/Numark-Mixtrack-3-scripts.js:
```diff
@@ -1984,138 +2119,245 @@ NumarkMixtrack3.PitchBendPlusButton = function(channel, control, value, status,
+                }
             }
+        } else {
+            for (i = 5; i <= 8; i++) {
```
You don't need duplicate code in an `if`...`else` here. Just use `(i = 1 + 4 * decknum; i <= 4 * decknum; i++)` for the loop.

djsteph replied:
Not sure I get it the magic of what you are teaching me... The if statement is related to which deck is the knob used from. What this is doing is if it is deck 1, then adjust gain of samplers 1-4, if deck 2, adjust gain of samplers 5-8. 
If I read properly, the variable "i" will always have the same value, so only the gain of samplers 1-4 would be affected, or all 8 samplers...

Be-ing replied:
No, the `i` variable will have a different value depending on the deck. Looking at this again, it would be clearer to define a variable so the `for` loop expression doesn't become so complicated:
```
var startingSampler = 1 + 4 * (decknum -1 );
for (i = startingSampler; i <= startingSampler + 4; i++) {
    ...
```

djsteph replied:
Thanks.. I think I got it... That made a huge chunk of code go away!

Be-ing commented on res/controllers/Numark-Mixtrack-3-scripts.js:
```diff
@@ -62,6 +56,13 @@ var intervalsPerRev = 1200,
+ * first 4 values used for Autoloop Not shifted
+ * last 4 values used for Autoloop Shifted
+ **************************/    
+var loopsize = [1, 2, 4, 8, 0.125, 0.25, 0.5, 16];
```
The jump from .5 to 16 seems kinda odd. That's really how Serato does it?

djsteph replied:
DJ Intro (Serato) and VDJ do not have Shift + Auto Loop values; they only have 1, 2, 4, 8. 
I picked 1/8, 1/4, 1/2 and 16  for shift as those were the other values available on the UI for the Shift values. My trial for Serato DJ (the full version) is expired so I can't confirm for that one. But here are the links to the documented mapping by each: 
http://www.virtualdj.com/documents/Numark%20Mixtrack%203%20-%20VirtualDJ%208%20Operation%20Guide.pdf and https://www.numark.com/images/product_downloads/MixtrackPro3-UserGuide-v1.1.pdf

Be-ing replied:
It may make more sense to break from their mappings for this. What do you think of 2, 4, 8, 16 beats unshifted and 1, 1/2, 1/4, 1/8 beats shifted?

djsteph replied:
The way I coded it makes it really easier for anyone to adjust as they wish in the config. PCDJ has it as 1/, 1, 4 and 16 not shifted so I tend to think that it is a matter of personal preference. I'm OK to commit as you say,

Be-ing commented on res/controllers/Numark-Mixtrack-3-scripts.js:
```diff
@@ -1984,138 +2119,245 @@ NumarkMixtrack3.PitchBendPlusButton = function(channel, control, value, status,
+        for (i = 1; i <= 3; i++) {
+            
+            if (deck.TapDown) {
+                gainValue[i-1] = engine.getParameter("[EffectRack1_EffectUnit" + [i] + "_Effect1]", "parameter2"); 
```
Why is `i` in brackets?

djsteph replied:
corrected... not a pro in javascript

djsteph committed:
Code cleanup with jsbeautifier and review comments incorporated

Clean up of code and review comments incorporated

djsteph commented:
Sorry for the large diff on the JS file. I don't believe I had used http://jsbeautifier.org/ the first time around, so a lot of minor changes related to spacing were made.

djsteph committed:
Exclude bpm.tapButton from JS Hint

Exclude bpm.tapButton from JS Hint

Be-ing commented:
> Sorry for the large diff on the JS file.

It helps to make a commit for each independent set of changes when that's practical. Graphical git clients can help with selecting which individual changes to include in a commit. On Windows you may want to try [Git For Windows](https://git-for-windows.github.io/) (I don't have any personal experience with it, but the screenshots look pretty good). You can make multiple commits offline before pushing them to GitHub.

djsteph committed:
Code cleanup in Sampler gain, changed Autoloop assigment

Code cleanup in Sampler gain, changed Autoloop assigment to [2, 4, 8,
16, 1, 0.5, 0.25, 0.125]

djsteph committed:
Bug fix - Icut feature and FaderStart

Crossfader movement was wrong for deck 2 as the decknum was not properly
passed to the function. variable "deckNum" replaced by "decknum" where
used to maintain consistency.

Fader start would stutter when moving volume fader. Added a variable to
only check the delta once instead of checking at every value change:
Stutter is now gone :)

djsteph commented:
Hello, 
I was just curious to know if all was OK to merge? Or if you were still expecting changes from me for this mapping? Someone offered to help with the 4 deck mapping of this mapping, and would like to have the latest map to start from. 
Thanks,
Steph

Be-ing commented:
I'll have time to resume review tomorrow or Friday.

Be-ing commented:
This has some interesting ideas for effects mappings, but I think some of it needs refinement. I'm primarily concerned with the user experience and how intuitive it is to use the controller with Mixxx. I'm only mentioning things I notice in the code as I reference it to understand the functionality. 

NumarkMixtrack3.StripTouchEffect and NumarkMixtrack3.StripTouchSearch seem to be the exact same function. Choose one name and delete the other.

The effects related functions are scattered widely throughout the JS file. Please organize them together into their own section.

> Tap button + Strip FX paramerer 1 - Effects 1, 2 and 3
> TAP + Beat knob = parameter 2 of FX 1, 2 and 3

Looking at [the code](https://github.com/djsteph/mixxx/blob/de837921bda32ab3adddd042b0e6ddfca590a3fb/res/controllers/Numark-Mixtrack-3-scripts.js#L1022), this seems to be adjusting parameter 1/2 of effect 1 in effect units 1, 2, and 3 simultaneously. This doesn't seem particularly useful considering that parameter 1 in one effect affects the sound in a completely different way from parameter 1 in another effect. How about only adjusting the lowest numbered effect unit linked to the deck with the FX1/2/3 buttons? For example, if a deck is linked to effect units 2 and 3, these mappings would only change parameter 1/2 of effect unit 2.

Similarly, InstantFX is adjusting the superknob of 3 effect units simultaneously. I think it should only adjust the superknob of the effect units linked to that deck.
It seems InstantFX toggles whether the effect units are enabled, but I think it would be more intuitive to enable the effects while the strip is held with a finger and disable them when the finger is lifted.

From your descriptions, it seems you're confusing individual effects with effect units, which makes me guess that you're using LateNight. In documentation, be sure to clearly distinguish between effect units and effects within effect units, or users of Deere will get quite confused.

Unfortunately, the way this is mapped doesn't allow much control over effect chains like Deere does. However, given the discrepancy between the skins currently, it has to be one way or the other (or add a whole other set of alternate mappings which will complicate the code & documentation and make it a pain to maintain).

djsteph commented:
Hello Be,

Thank you for taking the time to review and comment. 

> NumarkMixtrack3.StripTouchEffect and NumarkMixtrack3.StripTouchSearch seem to be the exact same function. Choose one name and delete the other.

The Midi code to trigger the function is different for both, so merging them would require that I add code to see which note is sent:

```
                <key>NumarkMixtrack3.StripTouchEffect</key>
                <midino>0x1A</midino>
                <key>NumarkMixtrack3.StripTouchSearch</key>
                <midino>0x1B</midino>`
```

I kept the following as this code was already there, I actually considered removing it completely..
```
Tap button + Strip FX paramerer 1 - Effects 1, 2 and 3
TAP + Beat knob = parameter 2 of FX 1, 2 and 3
```
Regarding the Effect section of the map: I use Dark Metal as my primary skin. I do test with Deere. My design decisions are based on the fact that the majority of skin (actually all but Deere) only expose one FX in each effect unit, or in other word, the first effect of a chain in a unit (do I make sense?). The way the FX Strip was mapped in the verison 1.3 of this mapping did not make sense in my opinion, as I can't see myself reaching to Deck 1 to control the FXs while on deck 2. I had not reworked it at that time for some of the reason you mentioned in your last paragraph. 

The way the initial mapping was (enabling of FXs (Simple tap on the FX and selection of FX (Shift + FX Button) worked fine so I did not modify that. 

I also largely based my decisions on the original mapping the intial mappers (although there is not much left from the original mapping), the one Virtual DJ made for this controller and also a bit on Serato. Granted FX implementation is a bit different in each one. I also used my experience (and preferences) as a mobile DJ in the past. As an example;  I like the flanger FX, Once it is set properly, I am fine to use that setting for both decks, I don't need to build a chain a many effects.

- Virtual DJ strip affects parameter 1 of all loaded FXs
- Serato DJ Intro strip affects ALL parameters of FX 1

The few comments I received regarding the strip, is that most people use it a a needle search, but the firmware forces the shift to be used for that in order to send midi note 0x1B. This is also how VDJ and Serato implemented the needle search (Shift+Strip)

If all skins were implemented the same way, it would be way easier to have a consensus. If only Deere were available, I would have mapped unit 1 to deck 1, with its FX 1, 2 and 3 and unit 2 to deck 2 with its own FXs 1, 2 and 3. as opposed to FX 1 = Unit 1, FX 2 = Unit 2, FX 3 = Unit 3 in the proposed map. Unfortunately, doing so, means that users of other skins will be blind for Unit 1, FX 2 and 3, and all of Unit 2, FX 1, 2, and 3. Not sure if this is a good move. 

> It seems InstantFX toggles whether the effect units are enabled,

Actually Instant FX is a 2 step process: 
1: User decides which effect are included in Instant FX by holding the TAP Button and the desired FX button.  This only adds them to an array in preparation for the using the strip.
2. While the strip is held with a finger, the FX included in Instant FX are enabled are disabled when the finger is lifted, just like you described.

`However, given the discrepancy between the skins currently, it has to be one way or the other (or add a whole other set of alternate mappings which will complicate the code & documentation and make it a pain to maintain).`

You are correct, that is the hard part, how to satisfy everyone with this current state. The only way currently would be to maintain a mapping for each skin, which is a not a good place to be.

djsteph committed:
Removed duplicate function for Strip

NumarkMixtrack3.StripTouchEffect and NumarkMixtrack3.StripTouchSearch
were identifical. NumarkMixtrack3.StripTouchSearch is removed in this
commit

Be-ing commented:
> You are correct, that is the hard part, how to satisfy everyone with this current state. The only way currently would be to maintain a mapping for each skin, which is a not a good place to be.

You can't fully satisfy everyone with the current state of the skins. I don't think there's a very good argument to choose one way over the other, it's just personal preference at this point.

> I also largely based my decisions on the original mapping the intial mappers (although there is not much left from the original mapping), the one Virtual DJ made for this controller and also a bit on Serato. 

I think this is a time to break from those mappings, as the effect system in Mixxx is not the same. In Mixxx, the superknob is meant for controlling multiple effects with a single control. This is already available through the InstantFX/touch strip. IMO it would make sense to use the tap + strip and tap + beats encoder mappings to have some limited control over individual effects. I realize this is also available with the alternate mappings of the EQ knobs, but those are all potentiometers with a finite range, so there's the awkwardness of soft takeover that doesn't exist with the touch strip and beats encoder.

djsteph commented:
Hello Be, 

Just to be sure what the expectation is, are you suggesting that the FX mapping be revised in order to accept this pull request?
To clarify the FX mapping, I made a table to detail what it currently is in this pull request, and what I think you expect: https://docs.google.com/spreadsheets/d/1JC-WTaYYQUPjHWUG99VQrJWkQVLrs1OtNzi4gZTNZzk/edit?usp=sharing

> it would make sense to use the tap + strip and tap + beats encoder mappings to have some limited control over individual effects

I am not sure where I would map the tap + strip and tap + beats encoder if I revise the mapping

djsteph commented:
I need guidance: I now have version 2.0 ready, which enables to use of 4 deck for this controller, has the FX mapped to perfection with Deere skin, has a few more features (e.g brake and spinback effects) and corrects a few bugs. In my opinion, this will be the final mapping for the MTP3. Should I cancel this pull request and submit v2.0 as a new one? or commit that version on top of this one? Or submit it along side this one? (I don't see the value to maintain 2 mappings for the same controller)?

Be-ing commented:
You could commit the new version to this branch and keep this PR open. I agree that it wouldn't make sense to maintain two mappings that are mostly the same.

djsteph committed:
Updated to version 2.0 - 4 deck capability added

Updated script with 4 deck capability, changed FX mapping to match Deere
skin, Brake and Spinback effects, improved Pad button handling and lots
code clean up.
http://mixxx.org/forums/viewtopic.php?f=7&t=7286&start=210#p30435

djsteph commented:
Done, Mixtrack Pro 3 mapping version 2.0 has been committed

djsteph committed:
Version 2.1 - Added FX Mode selection in setup variables

Allow use to choose FX Mode based on skin used:
FX Mode 1: (Deere skin): 1 Effect Unit with 3 effects per deck
FX Mode 2: (other skins): 1 effect available on 3 effect units, (1
effect unit per effect button on controller). FX settings are shared by
each deck.

Be-ing commented:
Is the wiki up to date with the latest mapping?

djsteph commented:
It is, I updated it about a month ago

Envoyé de mon iPhone

> Le 26 nov. 2016 à 17:06, Be <notifications@github.com> a écrit :
> 
> Is the wiki up to date with the latest mapping?
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

Be-ing commented:
It doesn't look like you have updated the wiki to document how the different effects mode options work. Could you do that?

djsteph commented:
The description of FXMode configuration is now on the wiki.

djsteph committed:
Wheel behavior improvement

Modified wheel behavior to always allow use of wheel to position song
when song is not playing
Added configuration option for to choose if pitch bend is allowed when
wheel is off

djsteph commented:
Wiki has been updated to reflect changes by the latest commit (	Wheel behavior improvement)
Thanks

radusuciu commented:
Thank you for this work, I've used these mappings and had no problems!

Be-ing commented:
Apologizes if this wasn't clear, but I'm holding off on reviewing this until the new effects interface in #1063 and #1062 is merged. If you'd like to contribute to those design discussions with how you think the Mixtrack Pro 3 could be mapped to the new system, please chime in.

djsteph commented:
Hello Be,
Not sure how the new effect interface would impact this. The only thing
that I see could change is the soft takevover fix for the parameters you
implemented, but I think the code I made to handle that could remain in the
mapping for backward compatibility, Other than that, the mapping already
manages the 2 different methods that are used for effects currently.

Stéphane Morin
Maison: 450 510-0622
Cellulaire: 514 243-5876

On Tue, Dec 20, 2016 at 1:05 PM, Be <notifications@github.com> wrote:

> Apologizes if this wasn't clear, but I'm holding off on reviewing this
> until the new effects interface in #1063
> <https://github.com/mixxxdj/mixxx/pull/1063> and #1062
> <https://github.com/mixxxdj/mixxx/pull/1062> is merged. If you'd like to
> contribute to those design discussions with how you think the Mixtrack Pro
> 3 could be mapped to the new system, please chime in.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/mixxxdj/mixxx/pull/1014#issuecomment-268313800>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AQDKya3iN1SJwt1Zo3_GYsV6V57z5XtUks5rKBj1gaJpZM4J_e6B>
> .
>

Be-ing commented:
The new effects interface introduces a new CO, [EffectUnitX_EffectY],meta that acts like a superknob for each effect in an EffectUnit. This will allow the mapping make use of effect chains.

djsteph commented:
I don't have plans to change this mapping further. I would prefer if it can
be incorporated in Github such that if someone want to help they will be
able to

Stéphane Morin
Maison: 450 510-0622
Cellulaire: 514 243-5876

On Tue, Dec 20, 2016 at 1:18 PM, Be <notifications@github.com> wrote:

> The new effects interface introduces a new CO, [EffectUnitX_EffectY],meta
> that acts like a superknob for each effect in an EffectUnit. This will
> allow the mapping make use of effect chains.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/mixxxdj/mixxx/pull/1014#issuecomment-268317023>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AQDKycNrjZX5YVhndY6Na4LsJeT4Sb4Cks5rKBwOgaJpZM4J_e6B>
> .
>

Be-ing commented:
@radusuciu , would you be interested in modifying this mapping for the new effects UI?

djsteph commented:
To honest, I am not sure how you would see the new configuration mapped on this controller. How would you map it? Currently each deck is mapped to its own effect unit and each fx button is mapped to its own effect in that specific unit.  The strip is mapped to the super button.

Be-ing commented:
Map the buttons to selecting [EffectUnitX],focused_effect. Map the touch strip to the meta knob for the focused effect. If no effect is focused, the touch strip controls the superknob. This mode could be entered by pressing the button for the focused effect when it is already focused. Map shift + the buttons to enabling/disabling each effect. I don't think this controller has enough physical controls for it to be useful to map different behavior depending on [EffectUnitX],show_parameters.

It may be helpful to switch around which behavior of the buttons is default and which is shifted. I feel it is annoying to have LEDs flicker when shift is pressed, so the LEDs should reflect either the focus selection or the effect enable switches, but not both. I think it will take trying both ways to decide which is better.

radusuciu commented:
I may be able to help with this, but wouldn't it be better to have those pull requests resolved and merged before starting to change up mappings?

Currently the effect buttons work such that pressing the button turns an effect on and off from a chain, with one chain per deck, and 4 effects per chain. Shift + effect button is mapped to selecting effects. This behaviour is also etched into the controller itself, and so it may be a bit confusing if re-mapped.

Be-ing commented:
> I may be able to help with this, but wouldn't it be better to have those pull requests resolved and merged before starting to change up mappings?

Yes, it would. Sorry if that wasn't clear.

> It may be helpful to switch around which behavior of the buttons is default and which is shifted.

On second thought, keeping the enable/disable switches as the default layer would probably be most useful.

> Currently the effect buttons work such that pressing the button turns an effect on and off from a chain, with one unit per deck, and 4 effects per unit. Shift + effect button is mapped to selecting effects. This behaviour is also etched into the controller itself, and so it may be a bit confusing if re-mapped.

It is okay to deviate from the labels on the controller a bit. It is nice to be able to scroll through the available effects from the controller, but generally that is something that is set up before using the effects, so it doesn't really need to be accessible from a controller.

Is the "beats" encoder a push encoder (does it send a signal when it is pressed)?

djsteph commented:
Beats knob is a infinite rotary knob that sends value 0 or 1 depending on the direction it is turned. It cannot be pushed.

radusuciu commented:
Isn't the new focus proposal very similar to the "Instant FX" feature as implemented for this controller? 

> Currently Tap + EffectButton adds effect to Instant FX so that when you touch the strip, it gets enabled and when you remove your finger, it 

> Tap + FX 1: Assign / unassign FX 1 to Instant FX. When assigned to Instant FX, the FX is instantly activated by touching the Strip and stopped when finger is lifted.

Be-ing commented:
> Isn't the new focus proposal very similar to the "Instant FX" feature as implemented for this controller?

Yes, but the current implementation of Instant FX does not allow control of individual metaknobs. It can only control them all together (through the superknob).

radusuciu commented:
Having just played with this a bit, the current implementation feels pretty nice as it turns off when you stop touching the strip. This could also be accomplished with the new effects rack setup but it would be a bit less smooth. Unless I misunderstand, you'd either have to swipe it down, or hit a button, or have some sort of configurable option.

djsteph commented:
> Yes, but the current implementation of Instant FX does not allow control of individual effects. It can only control them all together (through the superknob).

IMO, the only change that may make some sense, it to map the strip to active FXs Super knob instead of the effect unit super knob. But since there is only one strip to use with the 3 FX button, as soon as you enable more than one effect, you are in fact producing a very similar result as the current mapping produces with the Effect Unit super knob; so I see no real benefit to change this. If there were a knob per FX, then it would make sense to change the mapping, but this is not the case with this controller. 

> I feel it is annoying to have LEDs flicker when shift is pressed, so the LEDs should reflect either the focus selection or the effect enable switches, but not both. I think it will take trying both ways to decide which is better.

The only time LED will flash is if you activate InstantFX, so that you can tell that the effect is not simply activated. (Solid On). I did pay close attention to limit flashing LEDs to the strict essential. We normally have a light show on the dance floor, I don't need one in front of me on top of it :)

Be-ing commented:
Instant FX and controlling individual effects' metaknobs are mutually exclusive; the controller has to do one or the other at a time. I think I have come up with a way where either mode could be selected. Use tap button + turning the beats encoder to select the focused effect whose metaknob the touchstrip would control. If an effect is focused, turn off Instant FX. If the user scrolls up before the first effect or down past the last effect, revert to controlling the superknob. Instant FX would still be available the same way it is currently.

radusuciu commented:
I can commit to working to adapt this mapping for the new effect rack setup once that is merged. In the meantime, these updates can be merged imo as it all works quite well.

Be-ing commented:
Can someone merge this now? Adding support for the new effects can come in another PR.

daschuer commented:
Thank you for all the work!

Be-ing commented:
The new effects UI allowing better control of effect chains has been merged. So far only the Deere skin has been updated, but the other skins will be updated for 2.1 too. @djsteph and/or @radusuciu, would you like to contribute an updated mapping? You can find documentation about what has changed for mapping controllers on the [wiki](http://mixxx.org/wiki/doku.php/updating_controller_mappings).

radusuciu commented:
@Be-ing I'd be willing to contribute but would defer to @djsteph.

djsteph commented:
Hello Radu,

I don't have plans to modify the mapping. The latest version has been
merged in GIT, you can apply the modifications required. The temporary
functions for parameter soft takeover can be replaced with the proper
softtakeover. For the new effect setup, I'll let you decide the best use of
the controller.

Thanks
Steph

On Fri, Feb 3, 2017 at 4:27 PM, Radu Suciu <notifications@github.com> wrote:

> @Be-ing <https://github.com/Be-ing> I'd be willing to contribute but
> would defer to @djsteph <https://github.com/djsteph>.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/mixxxdj/mixxx/pull/1014#issuecomment-277366759>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AQDKycSU01murM7nusYM8BmddAu2xhvnks5rY5vNgaJpZM4J_e6B>
> .
>

Be-ing commented:
My general suggestion is to use the effect buttons in combination with some modifier, whether that's shift or the tap button or whatever, to set [EffectRack1_EffectUnitX],focused_effect. Pressing that again for the focused effect should set focused_effect back to 0. When an effect is focused, make the touch strip control that effect's metaknob; when no effect is focused, control the superknob as before. If the Instant FX feature is enabled, unfocus and control the superknob.

Be-ing commented:
I have a suggestion to simplify the mapping and use both Instant FX and the new effect focusing. Instead of assigning effects to Instant FX, make Instant FX mode a binary toggle switch. When Instant FX is on, touching the touchstrip would enable the focused effect and manipulate its metaknob. If no effect was focused when Instant FX is on, touching the touchstrip would enable any effects that are disabled and manipulate the superknob. This would allow the mapping for the effect buttons to be straightforward and remain similar to how they are now:

Press with no modifier: toggle effect enable switch
Press with Tap: focus/unfocus effect
Press with shift: load next effect

I'm not sure where the toggle switch for Instant FX mode would fit in. Maybe pressing and holding the Tap button?

radusuciu commented:
Will try and take a look at this tonight. Thanks for suggestions @Be-ing!

radusuciu commented:
I think I will use what Be outline above, including the Instant FX toggle with Tap long-press. I also agree that it would be most intuitive if focused effects map directly to Instant FX. When the Tap button is held, Instant FX mode will be toggled, and all focused effects will have their LEDs flash, the same as in the mapping now.

> If no effect was focused when Instant FX is on, touching the touchstrip would enable any effects that are disabled and manipulate the superknob. 

This part I think is less intuitive. If no effects are focused, and Instant FX is toggled on, do we just light up all of the available effect buttons? And if a user then focuses on an effect, do all the other ones turn off? I think I could get used to that but I'm not sure it's necessary. Currently, you have to make a deliberate choice to add an effect to Instant FX, and I think that that's fine.

Input appreciated, otherwise, will submit PR with above functionality.

By the way, I really like the way the new Effects rack looks/works in Deere. It feels a lot more practical.

Be-ing commented:
> This part I think is less intuitive. If no effects are focused, and Instant FX is toggled on, do we just light up all of the available effect buttons?

That might be a better idea than what I suggested.

> And if a user then focuses on an effect, do all the other ones turn off?

No, the point of focusing effects is not to have a bunch of effects to switch between. The point is to be able to chain effects and control them individually.

> By the way, I really like the way the new Effects rack looks/works in Deere. It feels a lot more practical.

:)

Be-ing commented:
If you were wondering, I don't think there would be any point in making the Mixtrack Pro 3 mapping work differently depending on the state of show_parameters. That is meant for controllers with more knobs.
